design/client_conformance_profiles.md

```diff
@@ -0,0 +1,550 @@
+In TLS handshake, the client is authenticated by its PKI X.509 Certificate.
+
+Keycloak binds this certificate with an access token and a refresh token by following Mutual-TLS Client Certificate-Bound Access Tokens [MTLS].
+
+* Reference:
+    - Authz Server / 5, 6
+    - Public Client / 1
+    - Confidential Client / 1
+
+Error Response : 400 Bad Request (the detail is TBD)
+
+#### Endpoint : Token Introspection Endpoint
+
+Client's Request Check
+
+* Operations :
+    - handle token introspection request
+
+##### 1. Client Authentication Method
+The client challenges its authentication by the method pre-registered in `token_endpoint_auth_method` of OpenID Connect Client Metadata (`tls_client_auth_subject_dn` or `private_key_jwt`).
+
+If `private_key_jwt`, the signature algorithm of JWS Client Assertion is `ES256`, `ES384`, `ES512`, `PS256`, `PS384` or `PS512`.
+
+* Reference:
+    - Authz Server / 14, 14-1, 14-2
+    - JWS algorithm considerations / 1, 2, 3
+
+Error Response : 400 Bad Request (the detail is TBD)
+
+#### Endpoint : Logout Endpoint
+
+Client's Request Check
+
+* Operations :
+    - handle (backchannel) logout request
+
+##### 1. Client Authentication Method
+The client challenges its authentication by the method pre-registered in `token_endpoint_auth_method` of OpenID Connect Client Metadata (`tls_client_auth_subject_dn` or `private_key_jwt`).
+
+If `private_key_jwt`, the signature algorithm of JWS Client Assertion is `ES256`, `ES384`, `ES512`, `PS256`, `PS384` or `PS512`.
+
+* Reference:
+    - Authz Server / 14, 14-1, 14-2
+    - JWS algorithm considerations / 1, 2, 3
+
+Error Response : 400 Bad Request (the detail is TBD)
+
+##### 2. Holder-of-Key Token
+In TLS handshake, the client is authenticated by its PKI X.509 Certificate.
+
+Keycloak verifies whether the refresh token is bound with this certificate by following Mutual-TLS Client Certificate-Bound Access Tokens [MTLS].
+
+* Reference:
+    - Authz Server / 5, 6
+    - Public Client / 1
+    - Confidential Client / 1
+
+Error Response : 400 Bad Request (the detail is TBD)
+
+#### Endpoint : UserInfo Endpoint
+
+Client's Request Check
+
+* Operations :
+    - handle UserInfo request
+
+##### 1. Holder-of-Key Token
+In TLS handshake, the client is authenticated by its PKI X.509 Certificate.
+
+Keycloak verifies whether the access token is bound with this certificate by following Mutual-TLS Client Certificate-Bound Access Tokens [MTLS].
+
+* Reference:
+    - Authz Server / 5, 6
+    - Public Client / 1
+    - Confidential Client / 1
+
+Error Response : 400 Bad Request (the detail is TBD)
+
+## TODO
+
+### "acr" (Authentication Context Class Reference) treatment
+
+FAPI security profile includes "acr" related items. However, keycloak has not yet determined what and when "acr" value is set.
+It has already been discussed in [KEYCLOAK-3314](https://issues.redhat.com/browse/KEYCLOAK-3314), but still open.
+
+### track update of FAPI Security Profile

```

Comment:
@tnorimat Shall we design the SPI (I'm favor of having a specific SPI for this) so that we can support different versions of the profiles?

Reply:
@pedroigor Yes, I agree with you. We prepare new SPI for this Client Conformance Profiles. 

My proposals is as follows :

* Implement the providers for the latest version (currently, implementer's draft 2 for both FAPI RO and FAPI RW).
* Additionally, implement the providers for the FINAL version when released and replace the previous providers with them.
* keycloak's UI does not show the version (e.g. FAPI RW Security Profile Implementer's Draft 2, FAPI RW Security Profile Implementer's Draft 3, FAPI RW Security Profile Final Draft), only show the name of Security Profile (e.g. FAPI RW Security Profile) due to the following reasons:
  - The slight differences will be expected between FAPI Security Profile's versions after current implementer's draft 2.
  - The user might not understand those slight differences between FAPI Security Profile's versions. If the version would be shown in UI, they might be confused.

WDYT?

Reply:
If companies are stuck with a specific version due to regulatory compliance (e.g.: regulatory framework not yet reflecting latest specs) I would say we should give them an option to choose a version.

You are probably in a better position than me to say whether or not this is a valid concern.

Reply:
Yes, I see as you said. So what about the following?

* keycloak's UI shows the profile including its version (e.g. FAPI RW Security Profile Implementer's Draft 2, FAPI RW Security Profile Implementer's Draft 3, FAPI RW Security Profile Final Draft) so that the admin can adhere to the specific version of the profile.